R-package/R/lgb.cv.R

```diff
@@ -333,6 +333,7 @@ lgb.cv <- function(params = list()
 
       booster <- Booster$new(params = params, train_set = dtrain)
       booster$add_valid(data = dtest, name = "valid")
+      booster$reset_parameter(params)
```

Comment:
note to reviewers: I don't totally understand *why* this change fixes #4667. At first I was worried that this function might suffer from a larger problem like "`lgb.cv()` does not respect values passed to `params`", but that is not true.

Consider the following example. Setting `min_gain_to_split` to a large number for the binary objective results in no splits, as expected.

```r
library(lightgbm)
data("agaricus.train")
dtrain <- lgb.Dataset(
    data = agaricus.train$data
    , label = agaricus.train$label
)
lgb.cv(
    params = list(
        min_gain_to_split = 100000.0
    )
    , data = dtrain
    , verbose = 0L
    , nrounds = 3L
    , nfold = 2L
    , obj = "binary"
)
```

The logs are full of the following

> [LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements

Changing to `min_gain_to_split = 0.0`, those warnings go away and training produces trees with multiple leaves.

Reply:
Maybe this related to the lack of corresponding code lines in R-package?
https://github.com/microsoft/LightGBM/blob/2ef3cb81ed66706ed12282857635f0a5941d92ad/python-package/lightgbm/engine.py#L222-L225

Reply:
Maybe? I'm not really sure. The code in `lgb.train()` and `lgb.cv()` in the R package is very similar, and `lgb.train()` doesn't have this issue with the parameter `verbose`.

Anyway, I think this fix is still worth merging even if it isn't fully understood...I don't think adding another `$update_params()` with the passed-in params introduces any risk.

Reply:
> I don't think adding another `$update_params()` with the passed-in params introduces any risk.

Agree. But this introduces performance degradation because this method calls C API function under the hood and is being executed for each fold.

Reply:
> but this introduces performance degradation

Sure, but I think that in this case, the cost of calling this method once per fold (one time at set up, not on each iteration, just to be clear) is worth it in exchange for ensuring that `lgb.cv()`'s behavior is correct.

Reply:
OK, I'm letting you to decide.

Reply:
Ok looking into this more closely tonight, I think I figured out:

1. why `verbose` is not being respected
2. why it seems like ONLY `verbose` is not being respected and all other parameters are working correctly

## Short Summary

Two bugs in the R package:

1. `Dataset$slice()` calls `LGBM_DatasetGetSubset()`, which will globally set the log level to INFO if `verbosity` isn't in the parameters passed to it. And `verbosity` is never passed through.
2. When `Dataset$update_params(new_params)` is called on a `Dataset` that has already been constructed, it doesn't store `new_params` in memory in the R object (`private$params`)

I just pushed https://github.com/microsoft/LightGBM/pull/4903/commits/619eaf484232642d9e231a5186a1d2cf2c48d74a which reverts the addition of `booster$reset_parameter()` and adds fixes for these two bugs.

## Much Longer Summary

I was able to reduce the internals of `lgb.cv()` to the following code that helps narrow down the issue.

<details><summary>sample code (click me)</summary>

```r
library(lightgbm)

data("agaricus.train", package = "lightgbm")

params <- list(
    max_bin = 123
    , verbose = -7L
)

dtrain <- lgb.Dataset(
    data = agaricus.train$data
    , label = agaricus.train$label
    , params = params
)
dtest <- slice(
    dataset = dtrain
    , idxset = seq_len(100L)
)

# with a validation set, you get logs
booster_w_valid <- lightgbm:::Booster$new(
    params = params
    , train_set = dtrain
)
booster_w_valid$add_valid(data = dtest, name = "valid")
booster_w_valid$update()

# without a validation set, no logs!
booster_no_valid <- lightgbm:::Booster$new(
    params = params
    , train_set = dtrain
)
booster_no_valid$update()
```

</details>

root cause: The use of `slice()` on a `Dataset` resets LightGBM's log level globally to the default value (INFO).

More details:

1. `Dataset$slice()` gets params from `Dataset$get_params()`
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L538
2. `Dataset$get_params()` filters down to only Dataset-specific parameters.
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L583-L592
3. `verbose` is not in the Dataset-specific parameters.
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/aliases.R#L9
4. `Dataset$slice()` eventually calls `Dataset$construct()` which calls `LGBM_DatasetGetSubset()`, which runs `config.Set()`
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/src/c_api.cpp#L1407
5. `config.Set()` calls `Log::ResetLogLevel()` which resets log level **globally** for the current process.
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/src/io/config.cpp#L241-L249
6. If `verbosity` isn't found in `params`, it defaults to INFO-level
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/include/LightGBM/config.h#L543

`Dataset.subset()` in the Python package doesn't suffer from this issue because it references `self.params` (the dictionary of parameters stored as an attribute on the `Dataset` instance) instead of using `Dataset.get_params()`, so `verbose` is not filtered out.

https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/python-package/lightgbm/basic.py#L1860

At first, I tried to just fix this by adding `"verbosity"` to the list of parameters returned by `Dataset$get_params()`. Unfortunately, that can lead to a new type of problem where the value of `verbosity` stored in a constructed Dataset will take precedence over values passed to `lgb.train()` / `lgb.cv()` if that Dataset is re-used multiple times.

https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Booster.R#L42

Next, I tried just changing the use of `Dataset$get_params()` to use `Dataset$private$params` in `Dataset$slice()`. That uncovered another bug. That still suffered from the problem that once the training Dataset was constructed, verbosity was locked in to the value from the last time it was constructed.

1. When calling `Dataset$construct()`, if it's already been constructed then R calls `LGBM_DatasetUpdateParamChecking()`
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L562
2. That call is try-catched. If it fails, then `Dataset$construct()` does the right thing based on whether or not raw data is still available. However, if that call does NOT fail, `private$params` is not updated!
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L559-L577